bagit.py

```diff
@@ -182,34 +214,73 @@ def make_bag(bag_dir, bag_info=None, processes=1, checksums=None, checksum=None,
                     continue
                 new_f = os.path.join(temp_data, f)
-                LOGGER.info(_('Moving %(source)s to %(destination)s'),
-                            {'source': f, 'destination': new_f})
-                os.rename(f, new_f)
+                if dest_dir:
+                    LOGGER.info(_('Copying %(source)s to %(destination)s'),
+                                {'source': os.path.join(bag_dir, f), 'destination': new_f})
+                    if os.path.isdir(os.path.join(bag_dir, f)):
+                        shutil.copytree(os.path.join(bag_dir, f), new_f)
+                    else:
+                        shutil.copy(os.path.join(bag_dir, f), new_f)
+                else:
+                    LOGGER.info(_('Moving %(source)s to %(destination)s'),
+                                {'source': os.path.join(bag_dir, f), 'destination': new_f})
+                    os.rename(os.path.join(bag_dir, f), new_f)
 
-            LOGGER.info(_('Moving %(source)s to %(destination)s'),
-                        {'source': temp_data, 'destination': 'data'})
-            os.rename(temp_data, 'data')
+            if data_subdir:
+                LOGGER.info(_('Moving %(source)s to %(destination)s'),
+                            {'source': temp_data, 
+                                'destination': os.path.join(data_dir, data_subdir)})
+                if not os.path.exists(data_dir):
+                    os.mkdir(data_dir)
+                elif os.path.exists(os.path.join(data_dir, data_subdir)):
+                    shutil.rmtree(os.path.join(data_dir, data_subdir))
+                os.rename(temp_data, os.path.join(data_dir, data_subdir))
+            else:
+                LOGGER.info(_('Moving %(source)s to %(destination)s'),
+                            {'source': temp_data, 'destination': data_dir})
+                os.rename(temp_data, data_dir)
 
             # permissions for the payload directory should match those of the
             # original directory
-            os.chmod('data', os.stat(cwd).st_mode)
+            os.chmod(data_dir, os.stat(bag_dir).st_mode)
+
+            if dest_dir:
+                total_bytes, total_files = make_manifests(
+                    data_dir, processes, algorithms=checksums, encoding=encoding, src_dir=bag_dir
+                )
+            else:
+                total_bytes, total_files = make_manifests(
+                    data_dir, processes, algorithms=checksums, encoding=encoding
+                )
 
-            total_bytes, total_files = make_manifests('data', processes, algorithms=checksums,
-                                                      encoding=encoding)
 
             LOGGER.info(_("Creating bagit.txt"))
             txt = """BagIt-Version: 0.97\nTag-File-Character-Encoding: UTF-8\n"""
-            with open_text_file('bagit.txt', 'w') as bagit_file:
-                bagit_file.write(txt)
+            if dest_dir:
```

Comment:
This entire section is where I was thinking it might be cleaner to always have `dest_dir` defined since I'm not happy about so many if statements in a row